File: Java/opentraceability/src/main/java/opentraceability/utility/DictionaryExtensions.java
package opentraceability.utility;

import java.util.HashMap;
import java.util.Map;

public class DictionaryExtensions {
    public static <TValue, TKey> Map<TValue, TKey> reverse(Map<TKey, TValue> source) {
        Map<TValue, TKey> dictionary = new HashMap<>();

        for (Map.Entry<TKey, TValue> entry : source.entrySet()) {
            if (!dictionary.containsKey(entry.getValue())){
                dictionary.put(entry.getValue(), entry.getKey());
            }
        }

        return dictionary;
    }
}

File: Kotlin/OpenTraceability/src/main/kotlin/opentraceability/utility/dictionaryextensions.kt
package opentraceability.utility

class DictionaryExtensions {
    companion object {
        fun<TValue, TKey> reverse(source: MutableMap<TKey, TValue>): MutableMap<TValue, TKey>{
            var dictionary: MutableMap<TValue, TKey> = mutableMapOf()

            source.forEach { entry ->
                if (!dictionary.containsKey(entry.value)){
                    dictionary.put(entry.value, entry.key);
                }
            }

            return dictionary
        }
    }
}

public fun <K, V> MutableMap<K, V>.reverse(): MutableMap<V, K> {
    return DictionaryExtensions.reverse(this)
}
